Sources/SwiftProtobuf/ProtobufBinarySizeVisitor.swift

```diff
@@ -0,0 +1,119 @@
+        let tagSize = Varint.encodedSize(of: UInt32(
+            truncatingBitPattern: protoFieldNumber << 3 | M.protobufWireType()))
+        let messageSize = try value.serializedSize()
+        serializedSize += tagSize + Varint.encodedSize(of: UInt64(messageSize)) + messageSize
+    }
+
+    mutating func visitRepeatedMessageField<M: ProtobufMessage>(value: [M], protoFieldNumber: Int, protoFieldName: String, jsonFieldName: String, swiftFieldName: String) throws {
+        let tagSize = Varint.encodedSize(of: UInt32(
+            truncatingBitPattern: protoFieldNumber << 3 | M.protobufWireType()))
+        serializedSize += value.count * tagSize
+        for v in value {
+            let messageSize = try v.serializedSize()
+            serializedSize += Varint.encodedSize(of: UInt64(messageSize)) + messageSize
+        }
+    }
+
+    mutating func visitSingularGroupField<G: ProtobufGroup>(value: G, protoFieldNumber: Int, protoFieldName: String, jsonFieldName: String, swiftFieldName: String) throws {
+        // The wire format doesn't matter here because the encoded size of the integer won't change
+        // based on the low three bits.
+        let tagSize = Varint.encodedSize(of: UInt32(truncatingBitPattern: protoFieldNumber << 3))
+        serializedSize += 2 * tagSize
+        try withAbstractVisitor {(visitor: inout ProtobufVisitor) in
+            try value.traverse(visitor: &visitor)
+        }
+    }
+
+    mutating func visitRepeatedGroupField<G: ProtobufGroup>(value: [G], protoFieldNumber: Int, protoFieldName: String, jsonFieldName: String, swiftFieldName: String) throws {
+        let tagSize = Varint.encodedSize(of: UInt32(truncatingBitPattern: protoFieldNumber << 3))
+        serializedSize += 2 * value.count * tagSize
+        for v in value {
+            try withAbstractVisitor {(visitor: inout ProtobufVisitor) in
+                try v.traverse(visitor: &visitor)
+            }
+        }
+    }
+
+    mutating func visitMapField<KeyType: ProtobufMapKeyType, ValueType: ProtobufMapValueType>(fieldType: ProtobufMap<KeyType, ValueType>.Type, value: ProtobufMap<KeyType, ValueType>.BaseType, protoFieldNumber: Int, protoFieldName: String, jsonFieldName: String, swiftFieldName: String) throws where KeyType.BaseType: Hashable {
+        let tagSize = Varint.encodedSize(of: UInt32(truncatingBitPattern: protoFieldNumber << 3))
+        let keyTagSize = Varint.encodedSize(of: UInt32(truncatingBitPattern: 1 << 3))
+        let valueTagSize = Varint.encodedSize(of: UInt32(truncatingBitPattern: 2 << 3))
+        for (k,v) in value {
+            let entrySize = try keyTagSize + KeyType.encodedSize(of: k) + valueTagSize + ValueType.encodedSize(of: v)
+            serializedSize += entrySize + Varint.encodedSize(of: Int64(entrySize))
+        }
+        serializedSize += value.count * tagSize
+    }
+}
```

Comment:
`encodedSize()` is an implementation detail; users don't see it. For each of the `ProtobufBinary*` types, it answers the question "how much space does this thing need on the wire after the tag is written", so in the case of messages it has to include that length prefix. The size visitor calls it abstractly to get the size of each field, whether it's an int, a string, a message, etc.